runtime/nvqir/QIRTypes.cpp

```diff
@@ -19,6 +20,37 @@ extern bool initialized;
 extern bool verbose;
 }
 
+nvqir::ArrayTracker &nvqir::ArrayTracker::getInstance() {
+  static thread_local nvqir::ArrayTracker instance;
+  return instance;
+}
+
+// Track the given array
+void nvqir::ArrayTracker::track(Array *arr) { allocated_arrays.push_back(arr); }
+// Untrack the given array (manually delete outside)
+void nvqir::ArrayTracker::untrack(Array *arr) {
+  auto it = std::find(allocated_arrays.begin(), allocated_arrays.end(), arr);
+  // Use nullptr to indicate untracked arrays (manually deleted outside) to
+  // prevent vector shrink.
+  if (it != allocated_arrays.end())
+    if (*it != nullptr)
+      *it = nullptr;
+    else
+      CUDAQ_WARN("Attempting to untrack an Array that is already untracked.");
```

Comment:
I think you could just test that `arr` was `nullptr` when passed in to get this warning condition.